src/graph/sampler.cc

```diff
@@ -524,4 +525,198 @@ IdArray SamplerOp::RandomWalk(
+      std::sort(neighbor_indices.begin(), neighbor_indices.end(), cmp);
+      for (const auto &pair : neighbor_indices) {
+        sub_indices->push_back(pair.first);
+        edge_mapping->push_back(pair.second);
+      }
+      sub_indptr->push_back(sub_indices->size());
+    }
+    flow_offsets->push_back(sub_indices->size());
+    first += src_size;
+  }
+  sub_eids->resize(sub_indices->size());
+  std::iota(sub_eids->begin(), sub_eids->end(), 0);
+}
+
+NodeFlow SamplerOp::LayerUniformSample(const ImmutableGraph *graph,
+                                       const IdArray seed_array,
+                                       const std::string &neighbor_type,
+                                       const std::vector<size_t> &layer_sizes) {
+  const auto g_csr = neighbor_type == "in" ? graph->GetInCSR() : graph->GetOutCSR();
+  const int64_t *indptr = g_csr->indptr.data();
+  const dgl_id_t *indices = g_csr->indices.data();
+  const dgl_id_t *eids = g_csr->edge_ids.data();
+
+  std::vector<dgl_id_t> layer_offsets;
+  std::vector<dgl_id_t> node_mapping;
+  std::vector<int64_t> actl_layer_sizes;
+  std::vector<float> probabilities;
+  ConstructLayers(indptr,
+                  indices,
+                  seed_array,
+                  layer_sizes,
+                  &layer_offsets,
+                  &node_mapping,
+                  &actl_layer_sizes,
+                  &probabilities);
+
+  std::vector<int64_t> sub_indptr;
+  std::vector<dgl_id_t> sub_indices;
+  std::vector<dgl_id_t> sub_eids;
+  std::vector<dgl_id_t> flow_offsets;
+  std::vector<dgl_id_t> edge_mapping;
+  ConstructFlows(indptr,
+                 indices,
+                 eids,
+                 node_mapping,
+                 actl_layer_sizes,
+                 &sub_indptr,
+                 &sub_indices,
+                 &sub_eids,
+                 &flow_offsets,
+                 &edge_mapping);
+
+  /*
+  std::cout << "layer_offsets:" << layer_offsets.size() << std::endl
+            << "node_mapping:" << node_mapping.size() << std::endl
+            << "actl_layer_sizes:" << actl_layer_sizes.size() << std::endl
+            << "probabilities:" << probabilities.size() << std::endl
+            << "sub_indptr:" << sub_indptr.size() << std::endl
+            << "sub_indices:" << sub_indices.size() << std::endl
+            << "sub_eids:" << sub_eids.size() << std::endl
+            << "flow_offsets:" << flow_offsets.size() << std::endl
+            << "edge_mapping:" << edge_mapping.size() << std::endl;
+  */
+
+  NodeFlow nf;
+
+  int64_t n_nodes = node_mapping.size();
+  int64_t n_edges = edge_mapping.size();
+  auto sub_csr = std::make_shared<ImmutableGraph::CSR>(n_nodes, n_edges);
+  std::copy(sub_indptr.begin(), sub_indptr.end(), sub_csr->indptr.begin());
+  std::copy(sub_indices.begin(), sub_indices.end(), std::back_inserter(sub_csr->indices));
+  std::copy(sub_eids.begin(), sub_eids.end(), std::back_inserter(sub_csr->edge_ids));
```

Comment:
potentially, you can give `sub_csr->indptr`, `sub_csr->indices` and `sub_csr->edge_ids` to `ConstructFlows` to avoid data copy.